pkg/operation/shoot/shoot.go

```diff
@@ -431,3 +436,51 @@ func ToNetworks(s *gardencorev1beta1.Shoot) (*Networks, error) {
 		APIServer: apiserver,
 	}, nil
 }
+
+func calculateContainerRuntimes(gardenClient client.Client, shoot *gardencorev1beta1.Shoot, seedNamespace string) (map[string]extensionsv1alpha1.ContainerRuntime, error) {
```

Comment:
Hm, why do we need these two functions here and the `ContainerRuntimesMap` at all? I guess you copied from the `Extension` resource, however, here the difference is that some `Extension` resources are globally enabled without the need to be in the `Shoot`. For `ContainerRuntimes`, however, all the desired `type`s are in the Shoot manifest, so we don't need this stuff here, I assume.

Reply:
Well, they do exist in shoot manifest, but for each worker separately, so there can be duplicated or unavailable. Here i basically get a distinct list of container runtimes and check for the registered controller registrations

Reply:
Yes, but why is this needed? I think you can simply, in the `DeployContainerRuntimes` function, loop over all worker pools and all container runtimes, and then create the `ContainerRuntime` objects in the seed. You don't need any `ControllerRegistration`s (this is to be checked elsewhere and out-of-scope of this PR). Or?

Reply:
Ok, so the ControllerRegistration check will be done instead in the:
botanist: RequiredExtensionsExist?

Reply:
> Ok, so the ControllerRegistration check will be done instead in the:
botanist: RequiredExtensionsExist?

Yes, exactly, that's enough.

> If i will remove this function i will have to calculate twice the set of container runtimes.
in the RequiredExtensionsExist and in the DeployContainerRuntimes. i can also change this function just to save a set of required container runtimes on the shoot. Than i will iterate over this set in both of the above functions, I indeed doesn't need the ControllerRegistration resource.

Hm, I don't understand. In the `RequiredExtensionExist` function you check whether a `ControllerRegistration` exists that supports the given `types` for the `ContainerRuntime` kind.
In the `DeployContainerRuntime` function you loop over each worker pool and then over each container runtime, and simply create the `ContainerRuntime` CRDs based on the provided `type` and `providerConfig`. Nothing to compute here. Yes, you have to loop in both functions, but that's how it's done for the other extension resources as well at the moment. I don't think it's a problem. :)